### PR TITLE
constexpr product

### PR DESCRIPTION
**Pull request by nicolasaunai**
This modifies the `core::product`  that is not constexpr and constexpr init of the `product(ConstArray<T,dim>(value)}` is useful

### CODE REVIEW

File: src/core/utilities/types.hpp [hunk 1]
```diff
@@ -232,7 +232,10 @@ namespace core
             return std::string{val};
         return std::nullopt;
     }
-    inline std::optional<std::string> get_env(std::string&& key) { return get_env(key); }
+    inline std::optional<std::string> get_env(std::string&& key)
```
Comment (nicolasaunai) [inline src/core/utilities/types.hpp:235]: this is formatting...

File: src/core/utilities/types.hpp [hunk 2]
```diff
@@ -241,11 +244,20 @@ namespace core
 namespace PHARE::core
 {
 template<typename Container, typename Multiplies = typename Container::value_type>
-Multiplies product(Container const& container, Multiplies mul = 1)
+constexpr Multiplies product(Container const& container, Multiplies init = 1)
```
Comment (PhilipDeegan) [inline src/core/utilities/types.hpp:247]: I think the point of this `Multiplies` type is to be able to override the return type

https://github.com/PhilipDeegan/PHARE/blob/mkn/src/core/utilities/types.hpp#L254

Comment (nicolasaunai) [reply src/core/utilities/types.hpp:247]: and the version I pushed breaks that?

Comment (PhilipDeegan) [reply src/core/utilities/types.hpp:247]: no, but I don't get the copy from `init` to `result`

Comment (nicolasaunai) [reply src/core/utilities/types.hpp:247]: ´Cause you’re too smart

**Review by PhilipDeegan — APPROVED** [review]

**Comment by nicolasaunai** [comment]
the question now is do we merge that or do we keep the master code and change to cpp20 to get the constexpr accumulate? Maybe we merge since accumulate is not on GPU anyway?

**Comment by nicolasaunai** [comment]
extracted from #740  and not useful for now in master..

**Pull request closed without merging**